src/native/credentials_provider.c

```diff
@@ -648,6 +659,258 @@ static int s_fill_in_logins(struct aws_array_list *logins, struct aws_byte_curso
     return AWS_OP_SUCCESS;
 }
 
+/*
+ * Optional auxiliary provider data used by the Cognito provider binding.  Keeps a reference to the
+ * CognitoLoginTokenSource java object that should be used to query dynamic login tokens before each
+ * HTTP request to Cognito.
+ *
+ * We ref count this just to be safe; every in-progress credentials query (technically there should only be at most 1,
+ * but to be paranoid we ignore that) keeps a reference to this object (in addition to the reference kept by the
+ * generic provider binding via the aux_data field).
+ */
+struct aws_login_token_source_data {
+    struct aws_allocator *allocator;
+    struct aws_ref_count ref_count;
+    JavaVM *jvm;
+    jobject login_token_source;
+};
+
+static void s_aws_login_token_source_data_on_zero_ref(void *user_data) {
+    struct aws_login_token_source_data *login_token_source_data = user_data;
+
+    AWS_FATAL_ASSERT(login_token_source_data != NULL);
+    JavaVM *jvm = login_token_source_data->jvm;
+
+    /********** JNI ENV ACQUIRE **********/
+    JNIEnv *env = aws_jni_acquire_thread_env(jvm);
+    if (env != NULL) {
+        if (login_token_source_data->login_token_source != NULL) {
+            (*env)->DeleteGlobalRef(env, login_token_source_data->login_token_source);
+        }
+    }
+
+    aws_jni_release_thread_env(jvm, env);
+    /********** JNI ENV RELEASE **********/
+
+    aws_mem_release(login_token_source_data->allocator, login_token_source_data);
+}
+
+static struct aws_login_token_source_data *s_aws_login_token_source_data_acquire(
+    struct aws_login_token_source_data *login_token_source_data) {
+    if (login_token_source_data != NULL) {
+        aws_ref_count_acquire(&login_token_source_data->ref_count);
+    }
+
+    return login_token_source_data;
+}
+
+static struct aws_login_token_source_data *s_aws_login_token_source_data_release(
```

Comment:
Trivial: Why have this function return anything at all?

Reply:
This was for the release, but as you explained 
``` my_struct = release(my_struct); ``` 
as a pattern to assign NULL to the thing being released makes sense.

Reply:
Oops, correct.